[css-shapes] Implement rect() shape

nt1m opened the pull request:
#### 60c71caf6ae8f78fcf776c5441ebdcd14f8cc5bf
<pre>
[css-shapes] Implement rect() shape
<a href="https://bugs.webkit.org/show_bug.cgi?id=261363">https://bugs.webkit.org/show_bug.cgi?id=261363</a>
rdar://114479534

Reviewed by Cameron McCormack.

Implement rect(top right bottom left [round ...]?) syntax.

<a href="https://drafts.csswg.org/css-shapes-1/#funcdef-basic-shape-rect">https://drafts.csswg.org/css-shapes-1/#funcdef-basic-shape-rect</a>

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-masking/animations/clip-path-interpolation-xywh-rect-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-masking/parsing/clip-path-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-masking/parsing/clip-path-valid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/motion/animation/offset-path-interpolation-006-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/motion/parsing/offset-path-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/motion/parsing/offset-path-parsing-valid-expected.txt:
* Source/WebCore/css/BasicShapeFunctions.cpp:
(WebCore::valueForBasicShape):
(WebCore::convertToLengthOrAuto):
(WebCore::basicShapeForValue):
* Source/WebCore/css/CSSBasicShapes.cpp:
(WebCore::buildRadiiString):
(WebCore::buildXywhString):
(WebCore::CSSRectShapeValue::CSSRectShapeValue):
(WebCore::CSSRectShapeValue::create):
(WebCore::CSSRectShapeValue::equals const):
(WebCore::buildRectString):
(WebCore::CSSRectShapeValue::customCSSText const):
(WebCore::buildInsetString):
* Source/WebCore/css/CSSBasicShapes.h:
(WebCore::CSSRectShapeValue::top const):
(WebCore::CSSRectShapeValue::right const):
(WebCore::CSSRectShapeValue::bottom const):
(WebCore::CSSRectShapeValue::left const):
(WebCore::CSSRectShapeValue::topLeftRadius const):
(WebCore::CSSRectShapeValue::topRightRadius const):
(WebCore::CSSRectShapeValue::bottomRightRadius const):
(WebCore::CSSRectShapeValue::bottomLeftRadius const):
* Source/WebCore/css/CSSValue.cpp:
(WebCore::CSSValue::visitDerived):
* Source/WebCore/css/CSSValue.h:
(WebCore::CSSValue::isRectShape const):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeBasicShapeRect):
(WebCore::CSSPropertyParserHelpers::consumeBasicShape):
* Source/WebCore/rendering/style/BasicShapes.cpp:
(WebCore::BasicShapeRect::create):
(WebCore::BasicShapeRect::BasicShapeRect):
(WebCore::BasicShapeRect::clone const):
(WebCore::BasicShapeRect::operator== const):
(WebCore::BasicShapeRect::path):
(WebCore::BasicShapeRect::canBlend const):
(WebCore::BasicShapeRect::blend const):
(WebCore::BasicShapeRect::dump const):
* Source/WebCore/rendering/style/BasicShapes.h:

Canonical link: <a href="https://commits.webkit.org/267863@main">https://commits.webkit.org/267863@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7dbdd25a461826731bf1eff0ef78b760e66daa56

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/17912 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/18238 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/18799 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/19740 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/16759 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/18108 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/21531 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/18394 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/18766 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/18128 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/18391 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/15578 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/20608 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/15618 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/16331 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/22865 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/16634 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/16500 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/20732 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/17062 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/14455 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/16163 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4262 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/20522 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/16910 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/rendering/style/BasicShapes.cpp:
```diff
@@ -323,6 +323,110 @@ void BasicShapeEllipse::dump(TextStream& ts) const
+    if (type() != other.type())
+        return false;
+
+    auto& otherXywh = downcast<BasicShapeRect>(other);
```
copypasta

nt1m replied:
Fixed

smfr commented on Source/WebCore/rendering/style/BasicShapes.h:
```diff
@@ -385,6 +386,59 @@ class BasicShapeInset final : public BasicShape {
+
+    void dump(TextStream&) const final;
+
+    Length m_top;
```
Use `RectEdges<>`?

nt1m replied:
Fixed

smfr commented on Source/WebCore/css/CSSBasicShapes.cpp:
```diff
@@ -300,6 +311,72 @@ String CSSXywhValue::customCSSText() const
 
+CSSRectShapeValue::CSSRectShapeValue(Ref<CSSValue> top, Ref<CSSValue> right, Ref<CSSValue> bottom, Ref<CSSValue> left, RefPtr<CSSValue> topLeftRadius, RefPtr<CSSValue> topRightRadius, RefPtr<CSSValue> bottomRightRadius, RefPtr<CSSValue> bottomLeftRadius)
+    : CSSValue(RectShapeClass)
+    , m_top(WTFMove(top))
```
If you WTFMove these arguments shouldn't they be passed as r-value references?

nt1m replied:
Fixed

heycam commented on Source/WebCore/css/CSSBasicShapes.cpp:
```diff
@@ -246,15 +246,13 @@ static bool buildRadii(Vector<String>& radii, const String& topLeftRadius, const
-static String buildXywhString(const String& insetX, const String& insetY, const String& width, const String& height,
-    const String& topLeftRadiusWidth, const String& topLeftRadiusHeight,
+
+// FIXME: Try to avoid using StringBuilder here.
```
Is avoiding it important for some reason?

nt1m replied:
It creates a lot of intermediate strings, but potentially the solution is to make StringBuilder stop doing that rather than avoiding using it (which is what's been done in many places)

darinadler replied:
It's slightly misleading to say that `StringBuilder` "creates additional strings". It does not do that. It creates a single result string. However, there are two best practices for efficiency:

- When it can be used to build the full result, `makeString` should be more efficient than `StringBuilder` because it computes the length up front and allocates once at the correct size, with no need to grow as we append and truncate when the string is complete
- When putting a string together from serialization of multiple things, using multiple `StringBuilder` objects means creating and destroying multiple buffers. Instead, building it up in a single `StringBuilder` by having functions that append to one avoids allocating and destroying incremental strings. This is not about avoiding `StringBuilder` but about avoiding multiple `StringBuilder` objects.

heycam commented on Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
```diff
@@ -6991,6 +6991,25 @@ static bool consumeShapeBorderRadius(CSSParserTokenRange& args, const CSSParserC
+        if (args.peek().id() == CSSValueAuto)
+            offset = consumeIdent<CSSValueAuto>(args);
+        else
+            offset = consumeLengthOrPercent(args, context.mode, ValueRange::All);
```
Maybe make `consumeAutoOrLengthOrPercent` accessible outside CSSPropertyParserHelpers.cpp and give it arguments you could use here?

heycam commented on Source/WebCore/rendering/style/BasicShapes.cpp:
```diff
@@ -323,6 +323,110 @@ void BasicShapeEllipse::dump(TextStream& ts) const
+
+Ref<BasicShape> BasicShapeRect::clone() const
+{
+    auto edges = m_edges;
+
+    auto topLeftRadius = m_topLeftRadius;
+    auto topRightRadius = m_topRightRadius;
+    auto bottomRightRadius = m_bottomRightRadius;
+    auto bottomLeftRadius = m_bottomLeftRadius;
+
+    return adoptRef(*new BasicShapeRect(WTFMove(edges), WTFMove(topLeftRadius), WTFMove(topRightRadius), WTFMove(bottomRightRadius), WTFMove(bottomLeftRadius)));
```
I think you could also write this as:

```
return adoptRef(*new BasicShapeRect(Ref { m_edges }, RefPtr { topLeftRadius }, ...));
```

and avoid the local variables.

nt1m replied:
The types don't match up (these are not RefPtr nor Ref)

heycam replied:
Aha, I was confusing it with the other class where they are Ref(Ptr)s.

heycam commented on Source/WebCore/rendering/style/BasicShapes.cpp:
```diff
@@ -323,6 +323,110 @@ void BasicShapeEllipse::dump(TextStream& ts) const
+        return false;
+
+    auto& otherRect = downcast<BasicShapeRect>(other);
+    return m_edges == otherRect.m_edges;
```
This is missing the radii comparisons.

nt1m replied:
Good catch!

heycam commented on Source/WebCore/rendering/style/BasicShapes.cpp:
```diff
@@ -323,6 +323,110 @@ void BasicShapeEllipse::dump(TextStream& ts) const
+    ts.dumpProperty("top-right-radius", topRightRadius());
+    ts.dumpProperty("bottom-right-radius", bottomRightRadius());
+    ts.dumpProperty("bottom-left-radius", bottomLeftRadius());
+
```
Nit: drop blank line.

heycam commented on Source/WebCore/rendering/style/BasicShapes.h:
```diff
@@ -385,6 +387,57 @@ class BasicShapeInset final : public BasicShape {
+    void setTopLeftRadius(LengthSize radius) { m_topLeftRadius = WTFMove(radius); }
+    void setTopRightRadius(LengthSize radius) { m_topRightRadius = WTFMove(radius); }
+    void setBottomRightRadius(LengthSize radius) { m_bottomRightRadius = WTFMove(radius); }
+    void setBottomLeftRadius(LengthSize radius) { m_bottomLeftRadius = WTFMove(radius); }
```
These could all be rvalue arguments.

heycam commented on Source/WebCore/rendering/style/BasicShapes.h:
```diff
@@ -385,6 +387,57 @@ class BasicShapeInset final : public BasicShape {
+
+    bool operator==(const BasicShape&) const override;
+
+    void dump(TextStream&) const final;
```
The class is final, so you may as well make all these say final instead of override.

heycam commented on Source/WebCore/rendering/style/BasicShapes.h:
```diff
@@ -407,10 +460,10 @@ class BasicShapeXywh final : public BasicShape {
     void setWidth(Length width) { m_width = WTFMove(width); }
     void setHeight(Length height) { m_height = WTFMove(height); }
```
These four functions could take rvalue arguments too.

nt1m committed:
[css-shapes] Implement rect() shape
https://bugs.webkit.org/show_bug.cgi?id=261363
rdar://114479534

Reviewed by Cameron McCormack.

Implement rect(top right bottom left [round ...]?) syntax.

https://drafts.csswg.org/css-shapes-1/#funcdef-basic-shape-rect

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-masking/animations/clip-path-interpolation-xywh-rect-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-masking/parsing/clip-path-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-masking/parsing/clip-path-valid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/motion/animation/offset-path-interpolation-006-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/motion/parsing/offset-path-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/motion/parsing/offset-path-parsing-valid-expected.txt:
* Source/WebCore/css/BasicShapeFunctions.cpp:
(WebCore::valueForBasicShape):
(WebCore::convertToLengthOrAuto):
(WebCore::basicShapeForValue):
* Source/WebCore/css/CSSBasicShapes.cpp:
(WebCore::buildRadiiString):
(WebCore::buildXywhString):
(WebCore::CSSRectShapeValue::CSSRectShapeValue):
(WebCore::CSSRectShapeValue::create):
(WebCore::CSSRectShapeValue::equals const):
(WebCore::buildRectString):
(WebCore::CSSRectShapeValue::customCSSText const):
(WebCore::buildInsetString):
* Source/WebCore/css/CSSBasicShapes.h:
(WebCore::CSSRectShapeValue::top const):
(WebCore::CSSRectShapeValue::right const):
(WebCore::CSSRectShapeValue::bottom const):
(WebCore::CSSRectShapeValue::left const):
(WebCore::CSSRectShapeValue::topLeftRadius const):
(WebCore::CSSRectShapeValue::topRightRadius const):
(WebCore::CSSRectShapeValue::bottomRightRadius const):
(WebCore::CSSRectShapeValue::bottomLeftRadius const):
* Source/WebCore/css/CSSValue.cpp:
(WebCore::CSSValue::visitDerived):
* Source/WebCore/css/CSSValue.h:
(WebCore::CSSValue::isRectShape const):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeBasicShapeRect):
(WebCore::CSSPropertyParserHelpers::consumeBasicShape):
* Source/WebCore/rendering/style/BasicShapes.cpp:
(WebCore::BasicShapeRect::create):
(WebCore::BasicShapeRect::BasicShapeRect):
(WebCore::BasicShapeRect::clone const):
(WebCore::BasicShapeRect::operator== const):
(WebCore::BasicShapeRect::path):
(WebCore::BasicShapeRect::canBlend const):
(WebCore::BasicShapeRect::blend const):
(WebCore::BasicShapeRect::dump const):
* Source/WebCore/rendering/style/BasicShapes.h:

Canonical link: https://commits.webkit.org/267863@main
